### ISSUE TITLE
lexer.peek() fails with error "not enough tokens available for peek operation"

### ISSUE THREAD

**Issue by ianb** [rse/tokenizr#20]
I feel like is so core that I must be misunderstanding, but lexer.peek() gives me error "not enough tokens available for peek operation" and the code seems to be clearly incorrect:

https://github.com/rse/tokenizr/blob/54807728b41ce992ac1fbc75b696c71da64a987e/src/tokenizr.js#L602-L611

```js
    peek (offset) {
        if (typeof offset === "undefined")
            offset = 0
        for (let i = 0; i < this._pending.length + offset; i++)
            this._tokenize()
        if (offset >= this._pending.length)
            throw new Error("not enough tokens available for peek operation")
        this._log(`PEEK: ${this._pending[offset].toString()}`)
        return this._pending[offset]
    }
```

So `lexer.peek()` leads to an `offset` of 0. If there's nothing in `this._pending` (the list of already parsed tokens) that code will resolve to:

```
for (let i=0; i < 0 + 0; i++)
    this._tokenize();
```

So that loop doesn't run at all. That means `this._tokenize()` won't be called, and `this._pending` won't have any tokens.

If I call `lexer._tokenize()` before `lexer.peek()` then everything works OK.

**Comment by gregzo** [comment]
I've run into the same issue. Because it is so core, my hunch is that it isn't meant to be called from within a rule function - all rules haven't run yet, so how can the lever know what the next token is?

My 2 cents: within a rule, use transactions and not `peek`.
When iterating over tokens (`Tokenizr.token()` generator), the cursor moves forward as tokens are consumed. There, `peek()` should be usable.

**Comment by ianb** [comment]
@gregzo I just made my own peek function:

```js
function peek(lexer) {
  if (!lexer._pending.length) {
    lexer._tokenize();
  }
  return lexer.peek();
}
```

This works, but does seem to be hungry(?), so `lexer.error()` doesn't give good file locations. (But it's better anyway when I keep track of a specific token to associate with errors.)

**Comment by rse** [comment]
This should be now fixed in version 1.7.0
